zebra-consensus/src/transaction/tests.rs

```diff
@@ -241,3 +241,50 @@ async fn v5_transaction_is_accepted_after_nu5_activation() {
         assert_eq!(result, Ok(expected_hash));
     }
 }
+
+/// Test if V5 transaction is rejected if script verifier returns an error.
+///
+/// This test simulates the case where the script verifier rejects the transaction by forcing it to
+/// return an error. To do so, a fake error is returned to the script verifier by the state service
+/// when an unspent transaction output is requested.
+#[tokio::test]
+// TODO: Remove `should_panic` once the NU5 activation heights for testnet and mainnet have been
+// defined.
+#[should_panic]
+async fn v5_transaction_is_rejected_based_on_script() {
+    let network = Network::Mainnet;
+    let blocks = zebra_test::vectors::MAINNET_BLOCKS.iter();
+
+    let state_service = service_fn(|_| async {
+        Err(Box::new(io::Error::new(
+            io::ErrorKind::Other,
+            "A fake error for the script verifier to forward to the transaction verifier",
+        )) as BoxError)
+    });
+
+    let script_verifier = script::Verifier::new(state_service);
+    let verifier = Verifier::new(network, script_verifier);
+
+    let transaction = fake_v5_transactions_for_network(network, blocks)
+        .rev()
+        .find(|transaction| {
+            !transaction.is_coinbase()
+                && transaction.inputs().is_empty()
+                && transaction.joinsplit_count() == 0
+                && transaction.sapling_spends_per_anchor().next().is_none()
+                && transaction.sapling_outputs().next().is_none()
+        })
+        .expect("At least one fake V5 coinbase transaction in the test vectors");
+
+    let result = verifier
+        .oneshot(Request::Block {
+            transaction: Arc::new(transaction),
+            known_utxos: Arc::new(HashMap::new()),
+            height: NetworkUpgrade::Nu5
+                .activation_height(network)
+                .expect("NU5 activation height is specified"),
+        })
+        .await;
+
+    assert!(result.is_err());
```

Comment:
This code doesn't actually check if a V5 transaction is rejected if the script verifier returns an error. Instead, it checks that the whole transaction verifier returns some kind of error for a fake v5 transaction.

Instead of just checking for any error, you could:
* check for the specific error string sent by the fake script verifier in `result.to_string()`
  * you might need to extract the error variant first, or downcast to `io::Error`
* run 3 tests, which together make sure the checks you want are reliable:
  * a successful v4 test vector with the real script verifier
  * a successful v4 test vector with the failing script verifier
  * a fake v5 transaction with the failing script verifier

I suggest just displaying the error.

Reply:
I decided to give it another try. I think I made it a bit less brittle and what is being tested is hopefully clearer and more reasonable. I also wrote passing and failing tests for V4 and V5. There's still an issue with V5 transaction tests because now they segfault. I guess that this is coming from the `zebra-script` FFI call, but I'm still investigating.